### PR TITLE
feat(AccessCode): Create and edit access codes

### LINKED ISSUE #276: Access code input box for code

### LINKED ISSUE #277: Generate code button

### PR DESCRIPTION
**Pull request by mikewuu**
closes #276 
closes #277 

### Notable things

- Did not implement code requirements (ie. numbers only, 4-8 characters) as this information could be different per device, and I don't think there's an endpoint to fetch it via seam-connect.
- Handled `invalid_access_code` error which contains more info on why a certain code failed to create (ie. reason for failing validation).
- Could **not** verify generating an access code. There seems to be a bug in `client.accessCodes.generateCode`, where the `device_id` is not being passed through as a query param.
- Made the **Generate code** button always visible vs. show on focus to avoid layout jumpiness (discussed this with @razor-x).

### Screenshots

image

#### Missing `device_id` when hitting `api/access_codes/generate_code` even though there are no TS errors.

CleanShot 2023-09-04 at 12 08 46

### CODE REVIEW

**Comment by razor-x** [comment]
Params should be fixed via https://github.com/seamapi/javascript/pull/277

**Comment by mikewuu** [comment]
> Params should be fixed via [seamapi/javascript#277](https://github.com/seamapi/javascript/pull/277)

ah, thank you for the quick fix!

**Comment by razor-x** [comment]
I am not suggesting adding this right now, but we may want to consider bringing in https://www.react-hook-form.com/ if our forms / field level validation gets more complex.

**Comment by razor-x** [comment]
> Did not implement code requirements (ie. numbers only, 4-8 characters) as this information could be different per device, and I don't think there's an endpoint to fetch it via seam-connect.

We can use https://docs.seam.co/latest/products/smart-locks/access-codes/access-code-requirements-for-door-locks#supported-code-lengths

Issue here: https://github.com/seamapi/react/issues/436

**Comment by mikewuu** [comment]
@razor-x good to merge this, and handle https://github.com/seamapi/react/issues/436 in a separate PR?

**Comment by razor-x** [comment]
Yes, but can we make sure generate code works before merge?

**Review by razor-x — APPROVED** [review]

**Comment by mikewuu** [comment]
> Yes, but can we make sure generate code works before merge?

Sure against the fake connect, or live?

File: src/lib/seam/access-codes/use-update-access-code.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ import type {
   AccessCodeUpdateRequest,
   AccessCodeUpdateResponse,
   ActionAttempt,
-  SeamError,
+  SeamAPIError,
```
[the remark this answers is not in the document]
Comment (mikewuu) [reply src/lib/seam/access-codes/use-update-access-code.ts:10]: I think we should, but let me verify that this is actually the correct type first. Just a best guess at the moment :)

**Comment by razor-x** [comment]
We can test it works for both.

**Comment by mikewuu** [comment]
@razor-x 

works 🙌 !! 

### Fake

https://github.com/seamapi/react/assets/11449462/d4bc248a-be83-4fbf-b66f-f33b47a5992f

### Real 


https://github.com/seamapi/react/assets/11449462/c5f8ad3b-faf3-4642-8206-3467976b9c70

File: src/lib/seam/access-codes/use-update-access-code.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikewuu) [reply src/lib/seam/access-codes/use-update-access-code.ts:10]: @razor-x  couldn't validate, I think we need to whitelist /create endpoint for CST at connect?

image

[the remark this answers is not in the document]
Comment (mikewuu) [reply src/lib/seam/access-codes/use-update-access-code.ts:10]: PR for this https://github.com/seamapi/seam-connect/pull/4226

[the remark this answers is not in the document]
Comment (mikewuu) [reply src/lib/seam/access-codes/use-update-access-code.ts]: @razor-x Fix looks good 👍 

### Create code working

https://github.com/seamapi/react/assets/11449462/b288e691-2eb7-45df-9453-54527bd02bf0

### Updated error parsing from https://github.com/seamapi/react/pull/428/commits/0d08a98555302804b56db3262938c2f5a761541a works

https://github.com/seamapi/react/assets/11449462/482d4757-12bf-4b95-afaf-675c1ebb5345

File: src/lib/seam/components/DeviceDetails/LockDeviceDetails.tsx [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import classNames from 'classnames'
 import type { LockDevice } from 'seamapi'
 
+import type { CommonProps } from 'lib/index.js'
```
Comment (razor-x) [inline src/lib/seam/components/DeviceDetails/LockDeviceDetails.tsx:4]:
```suggestion
import type { CommonProps } from 'lib/seam/components/common-props.js'
```

**Review by razor-x — APPROVED** [review]

**Pull request merged**